docs/src/references/mqtt-topic-guidelines.md

```diff
@@ -0,0 +1,827 @@
+```
+
+The `Rpi1001` device must be registered as a `device` by publishing the following retained message:
+
+```sh te2mqtt
+tedge mqtt pub -r te/_/Rpi1001 '{
+  "type": "device"
+}'
+```
+
+The device is registered under the default namespace indicated by `_` at the second level.
+The registration message supports additional fields like `name`, `parent`, `device-type` etc as well,
+which will be explained in detail later.
+
+Once the device is registered, the `tedge-agent` service can be registered as a `service` type as follows:
+
+```sh te2mqtt
+tedge mqtt pub -r te/_/Rpi1001/_/tedge-agent '{
+  "type": "service"
+}'
+```
+
+Once these entities are registered, they can start sending data or receive commands.
+
+Examples:
+
+* Device measurement:
+
+   ```sh te2mqtt
+   tedge mqtt pub -r te/_/Rpi1001/_/_/mes/battery_reading` '{
+      "charge": 88
+      "temperature": 32,
+      "voltage": 45,
+      "current": 15
+   }'
+   ```
+
+* Service measurement
+
+   ```sh te2mqtt
+   tedge mqtt pub -r te/_/Rpi1001/_/tedge-agent/mes/cpu_usage` '{
+      "usage": 23
+      "threads": 36,
+      "up_time": 3726,
+   }'
+   ```
+
+If the the user had used a different entity identification scheme like
+`te/<device-family>/<device-series>/<part-series>/<part-id>/...` instead,
+where the `<part-id>` is the unique device identifier,
+the registration message should have to be sent to the ``te/<device-family>/<device-series>/<part-series>/<part-id>` topic,
+with the `type` as `device`.
+
+So, it is completely up-to the user to choose how many levels he wants to uniquely identify devices and services.
+
+If your use-case doesn't need so many levels to identify an entity,
+the unnecessary levels can be skipped with the `_` (underscore) character as follows:
+
+```
+te/<entity-id>/_/_/_/...
+```
+
+or 
+
+```
+te/_/_/_/<entity-id>/...
+```
+
+**Main Device**
+
+The main device does not need any explicit registration and can be referred to using the `main` alias as device id.
+Hence, a measurement associated to the main device can be sent to following topic:
+
+```
+te/_/main/_/_/mes/battery_reading
+```
+
+But if some additional metadata needs to be added to the main device,
+they can be pushed to the `te/_/main` topic as done using a registration message.
+
+**Immediate Child Devices**
+
+Immediate child devices of the main device can be registered using the registration protocol described above.
+If a device is registered with `main` as the `parent` field value, or without a `parent` field,
+it is assumed to be an immediate child device.
+
+**Nested Child Devices**
+
+Nested child devices must always register explicitly by declaring their `parent`.
+
+### Automatic registration
+
+If an explicit registration is not done, thin-edge will assume the following topic level convention:
+
+```
+te/<device-namespace>/<device-id>/<service-namespace>/<service-id>
+```
+
+and auto-register the entities as per their positions in the topics.
```

Comment:
There is a risk with this auto-registration mechanism.

Even if a user explicitly sends a registration message first, before sending any telemetry data with that topic scheme, there is the possibility of `tedge` getting the telemetry data message first before the registration message, as there is no ordering guarantees across topics. In that case, when `tedge` get the first telemetry message, it will assume the default namespace scheme, as it hasn't seen a registration message yet, and will process that message with that default scheme. Then when the actual registration message is received, it might conflict with the default scheme (may be the 2nd topic level is not the device id but the 4th one).

To avoid such issues, we'll have to delay the processing of the auto-registration to give enough time for the registration messages to arrive, if any.

Reply:
A note on being able to turn off the auto registration has been added, and auto registration is limited to `te/device/+/service/+`